src/server/snap.rs

```diff
@@ -0,0 +1,196 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{self, Formatter, Display};
+use std::io;
+use std::fs::File;
+use std::net::{SocketAddr, TcpStream};
+use std::io::{Read, Write};
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::boxed::FnBox;
+use std::sync::{Arc, RwLock};
+use std::time::Instant;
+use threadpool::ThreadPool;
+use mio::Token;
+use bytes::{Buf, ByteBuf};
+use protobuf::Message;
+
+use super::{Result, ConnData};
+use super::transport::RaftStoreRouter;
+use raftstore::store::{SNAP_GEN_PREFIX, SNAP_REV_PREFIX, SnapFile};
+use util::worker::Runnable;
+use util::codec::rpc;
+use util::codec::number::NumberEncoder;
+use util::HandyRwLock;
+
+use kvproto::raft_serverpb::RaftMessage;
+
+pub type Callback = Box<FnBox(Result<()>) + Send>;
+
+const DEFAULT_SENDER_POOL_SIZE: usize = 3;
+
+/// `Task` that `Runner` can handle.
+///
+/// `Register` register a pending snapshot file with token;
+/// `Write` write data to snapshot file;
+/// `Close` save the snapshot file;
+/// `Discard` discard all the unsaved changes made to snapshot file;
+/// `SendTo` send the snapshot file to specified address.
+pub enum Task {
+    Register(Token, RaftMessage),
+    Write(Token, ByteBuf),
+    Close(Token),
+    Discard(Token),
+    SendTo {
+        addr: SocketAddr,
+        data: ConnData,
+        cb: Callback,
+    },
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            Task::Register(token, ref meta) => write!(f, "Register {:?} token: {:?}", meta, token),
+            Task::Write(token, _) => write!(f, "Write snap for {:?}", token),
+            Task::Close(token) => write!(f, "Close file {:?}", token),
+            Task::Discard(token) => write!(f, "Discard file {:?}", token),
+            Task::SendTo { ref addr, ref data, .. } => {
+                write!(f, "SendTo Snap[to: {}, snap: {:?}]", addr, data.msg)
+            }
+        }
+    }
+}
+
+/// Send the snapshot to specified address.
+///
+/// It will first send the normal raft snapshot message and then send the snapshot file.
+fn send_snap(snap_dir: PathBuf, addr: SocketAddr, data: ConnData) -> Result<()> {
+    assert!(data.is_snapshot());
+    let timer = Instant::now();
+    let snap = data.msg.get_raft().get_message().get_snapshot();
+    let mut snap_file = box_try!(SnapFile::from_snap(snap_dir.as_path(), &snap, SNAP_GEN_PREFIX));
+    if !snap_file.exists() {
+        return Err(box_err!("missing snap file: {:?}", snap_file.path()));
+    }
+    // snapshot file has been validated when created, so no need to validate again.
+
+    let mut f = try!(File::open(snap_file.path()));
+    snap_file.delete_when_drop();
```

Comment:
Actually it's not a setter. It's just a function which command the snap_file to delete itself when it's dropped.